src/scheduler/MpiWorld.cpp

```diff
@@ -25,19 +19,84 @@ MpiWorld::MpiWorld()
   , cartProcsPerDim(2)
 {}
 
-faabric::scheduler::FunctionCallClient& MpiWorld::getFunctionCallClient(
-  const std::string& otherHost)
+void MpiWorld::sendRemoteMpiMessage(
+  int sendRank,
+  int recvRank,
+  const std::shared_ptr<faabric::MPIMessage>& msg)
 {
-    auto it = functionCallClients.find(otherHost);
-    if (it == functionCallClients.end()) {
-        // The second argument is forwarded to the client's constructor
-        auto _it = functionCallClients.try_emplace(otherHost, otherHost);
-        if (!_it.second) {
-            throw std::runtime_error("Error inserting remote endpoint");
+    // Assert the ranks are sane
+    assert(0 <= sendRank && sendRank < size);
+    assert(0 <= recvRank && recvRank < size);
+
+    // Initialise the endpoint vector if not initialised
+    if (mpiMessageEndpoints.size() == 0) {
+        for (int i = 0; i < size * size; i++) {
+            mpiMessageEndpoints.emplace_back(nullptr);
         }
```

Comment:
I can see you've ported these changes into the other PR which I reviewed accidentally (https://github.com/faasm/faabric/pull/111), so the comments apply to both. 

Relevant comments from the other PR:

a) Rather than spread it around, I think lazy initialisation for the rank should all happen in one place.
b) I think you can do `mpiMessageEndpoints = std::vector(size * size, nullptr)` and save a few lines.

Reply:
b) `mpiMessageEndpoints` is a vector of `std::unique_ptr` so it has a deleted `=` operator. Thus I need to emplace `nullptr`s back. Note that this also answers your concern in #111 wrt `mpiMessageEndpoint[index` being actually initialised to `nullptr`.

Reply:
Ah that's annoying, makes sense though.